### ISSUE TITLE
Add default error codes

### ISSUE THREAD

**Issue by eahefnawy** [serverless/serverless#608]
Thanks to the awesome @staceymoore, we should now have the following default error codes templates in all projects `s-templates.json` file:

```
"responsesForGet": {
    ".*400.*": "$${response400}",
    ".*Process exited.*": "$${response500}",
    ".*Task timed out.*": "$${response504}",
    "default": "$${response200}"
  },
  "responsesForPost": {
    ".*400.*": "$${response400}",
    ".*Process exited.*": "$${response500}",
    ".*Task timed out.*": "$${response504}",
    "default": "$${response201}"
  },
  "response200": {
    "statusCode": "200",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": {
      "application/json": ""
    }
  },
  "response201": {
    "statusCode": "201",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": {
      "application/json": ""
    }
  },
  "response400": {
    "statusCode": "400",
    "responseModels": {},
    "responseTemplates": {
      "application/json": "#set($inputRoot = $input.path('$.errorMessage'))$inputRoot"
    }
  },
  "response500": {
    "statusCode": "500",
    "responseTemplates": {
      "application/json": "{\"error\": \"Server Error\"}"
    }
  },
  "response504": {
    "statusCode": "504",
    "responseTemplates": {
      "application/json": "{\"error\": \"Gateway Timeout\"}"
    }
  },
  "responseParameters": {
    "method.response.header.location": "integration.response.body.links.self"
  }
}
```

**Comment by staceymoore** [comment]
I'll add some notes and improvements tomorrow. Right now I need sleep. :-)

**Comment by ac360** [comment]
:+1:

**Comment by staceymoore** [comment]
A few notes for reference:

- To use something other than 200 as the success response for an endpoint, just set the default to that response, e.g. use `response201` as the default for POST endpoints as shown
- To set a Location header include the mapping in `responseParameters` (note that `integration.response.body.links.self` references a link that you would need to add to your success response object)
- To return a custom JSON error response, follow the example shown for response400 and do the following in your code:
  - Create an `error` object and include the the response code you want to use, e.g. `status: 400`
  - Call `content.done( JSON.stringify(error) )` (if you are using the default handlers in Serverless, you can have the handler stringify the error that is passed in with the callback)
  - Include a regexp in your responses template that will catch the response code, e.g. `".*400.*"`
  - Note that the mapping template `#set($inputRoot = $input.path('$.errorMessage'))$inputRoot` is what returns the proper JSON, if you leave it set to passthrough you will only get an escaped string
- `response500` catches early exit errors thrown by Lamba; customize the response as you wish, the only requirement is that it be a string in the template
- `response504` catches Lambda timeouts; customize the response as you wish, the only requirement is that it be a string in the template

I will fine tune these as I am working on my app over the next few days and will share any updates and improvements.

**Comment by eahefnawy** [comment]
Thanks a lot @staceymoore :blush: ... these are all super helpful!

**Comment by eahefnawy** [comment]
@staceymoore could you also share how your `s-function.json` looks like using these templates? :blush:

**Comment by staceymoore** [comment]
@eahefnawy  Note that `responsesForGet` and `responsesForPost` are parent templates that reference the other child templates. In the s-function.json I just refer to the parent templates, so it is very short. Here's an example:

```
"endpoints": [
    {
      "path": "merchants",
      "method": "POST",
      "type": "AWS",
      "authorizationType": "none",
      "apiKeyRequired": true,
      "requestParameters": {},
      "requestTemplates": "$${apiRequestTemplate}",
      "responses": "$${responsesForPost}"
    }
  ]
```

**Comment by staceymoore** [comment]
@eahefnawy and since it isn't included with the error templates above, below is what my request template currently looks like. I'm probably going to refactor a little to organize it by groups: context, header, path, querystring, body.

```
"apiRequestTemplate": {
    "application/json": {
      "resourcePath": "$context.resourcePath",
      "stage": "$context.stage",
      "httpMethod": "$context.httpMethod",
      "apiKey": "$input.params().header.get('x-api-key')",
      "sourceIp": "$input.params().header.get('X-Forwarded-For')",
      "auth": "$input.params().header.get('Authorization')",
      "path": {
        "all": "$input.params().path",
        "id": "$input.params().path.get('id')"
      },
      "query": {
        "all": "$input.params().querystring"
      },
      "body": "$input.json('$')"
    }
  }
```

**Comment by martinmicunda** [comment]
I am trying to pass `operation` value through `requestTemplates` to handler but it doesn't seem to work. Am I missing something or this is a bug? I also tried @staceymoore example above but event in handler is always empty.

Note: I am using serverless v0.4.2.

`s-function.json`
```json
  "endpoints": [
    {
      "path": "locations",
      "method": "GET",
      "type": "AWS",
      "authorizationType": "none",
      "apiKeyRequired": false,
      "requestParameters": {},
      "requestTemplates": {
        "application/json": "{\"operation\":\"list\"}"
      },
      "responses": {
        "400": {
          "statusCode": "400"
        },
        "default": {
          "statusCode": "200",
          "responseParameters": {},
          "responseModels": {},
          "responseTemplates": {
            "application/json": ""
          }
        }
      }
    },
    {
      "path": "locations/{id}",
      "method": "GET",
      "type": "AWS",
      "authorizationType": "none",
      "apiKeyRequired": false,
      "requestParameters": {},
      "requestTemplates": {
        "application/json": "{\"operation\":\"read\",\"id\":\"$input.params('id')\"}"
      },
      "responses": {
        "400": {
          "statusCode": "400"
        },
        "default": {
          "statusCode": "200",
          "responseParameters": {},
          "responseModels": {},
          "responseTemplates": {
            "application/json": ""
          }
        }
      }
    }
  ]
```

`handler.js`
```
module.exports.handler = function(event, context) {
    console.log('Received event:', JSON.stringify(event, null, 2));
};
```
**output:** `Received event: {}`

**Comment by staceymoore** [comment]
@martinmicunda tare you running the function locally? If so, make sure you have the event data in event.json inside your function's folder. By default it is an empty event, my guess is that's what you are seeing.

**Comment by erikerikson** [comment]
The following is our [ridiculous?] conclusion (altered slightly to anonymize and generalize).  Basically, we define reasonable default values at the root of the project for all API resource/method combinations and then override them with the correct specific values at the level of the actual function that is being defined.  Commentary at end.

Warning: I am still mid-migration from legacy versions of SLS so this may or may not function as expected.  I assume @staceymoore 's code is working nicely at it would seem to and trying to use the same pattern of recursive template inclusion.

At the s-project.json level:
```
{
  "functionTemplate": {
    "name": "$${functionName}",
    "customName": "$${functionNameTemplate}",
    "customRole": "$${roleNameTemplate}",
    "handler": "$${handlerTemplate}",
    "timeout": "$${timeout}",
    "memorySize": "$${memorySize}",
    "custom": "$${customTemplate}",
    "endpoints": "$${endpointsTemplate}",
    "events": "$${eventsTemplate}"
  },

  "functionName": "[OVERRIDE_ME_OR_ERROR]",
  "functionNameTemplate": "${project}-${stage}-${name}",
  "roleNameTemplate": "${project}-${stage}-${name}",
  "handlerTemplate": "${name}/handler.handler", // with multiple functions in s-function.json, would add "handlerGetTemplate", "handlerPostTemplate", et cetera
  "timeout": 3,
  "memorySize": 1024,
  "customTemplate": {
    "excludePatterns": [],
    "envVars": [ "SERVERLESS_STAGE" ]
  },
  "endpointsTemplate": [],
  "eventsTemplate": [],

  "endpointGetTemplate": {
    "path": "$${pathTemplate}",
    "method": "GET",
    "type": "$${typeTemplate}",
    "authorizationType": "$${authorizationTypeTemplate}",
    "apiKeyRequired": "$${apiKeyRequiredTemplate},
    "requestParameters": "$${requestParametersTemplate}",
    "requestTemplates": "$${requestTemplatesTemplate}",
    "responses": "$${responsesGetTemplate}"
  },
  "endpointPatchTemplate": {
    "path": "$${pathTemplate}",
    "method": "PATCH",
    "type": "$${typeTemplate}",
    "authorizationType": "$${authorizationTypeTemplate}",
    "apiKeyRequired": "$${apiKeyRequiredTemplate}",
    "requestParameters": "$${requestParametersTemplate}",
    "requestTemplates": "$${requestTemplatesTemplate}",
    "responses": "$${responsesPatchTemplate}"
  },
  "endpointPostTemplate": {
    "path": "$${pathTemplate}",
    "method": "POST",
    "type": "$${typeTemplate}",
    "authorizationType": "$${authorizationTypeTemplate}",
    "apiKeyRequired": "$${apiKeyRequiredTemplate}",
    "requestParameters": "$${requestParametersTemplate}",
    "requestTemplates": "$${requestTemplatesTemplate}",
    "responses": "$${responsesPostTemplate}"
  },
  
  "pathTemplate": "${component}/{id}/${name}",
  "typeTemplate": "AWS",
  "authorizationTypeTemplate": "AWS_IAM",
  "requestParametersTemplate": {},
  "requestTemplatesTemplate": { "application/json": "" },

  "responsesGetTemplate": {
    "default": "$${response200}",
    "400:.*":  "$${response400}",
    "403:.*":  "$${response403}",
    "404:.*":  "$${response404}",
    "422:.*":  "$${response422}",
    "500:.*":  "$${response500}",
    "504:.*":  "$${response504}"
  },
  "responsesPatchTemplate": {
    "default": "$${response204}",
    "400:.*":  "$${response400}",
    "403:.*":  "$${response403}",
    "404:.*":  "$${response404}",
    "409:.*":  "$${response409}",
    "412:.*":  "$${response412}",
    "422:.*":  "$${response422}",
    "500:.*":  "$${response500}",
    "504:.*":  "$${response504}"
  },
  "responsesPostTemplate": {
    "default": "$${response201}",
    "400:.*":  "$${response400}",
    "403:.*":  "$${response403}",
    "422:.*":  "$${response422}",
    "500:.*":  "$${response500}",
    "504:.*":  "$${response504}"
  },

  "response200": {
    "statusCode": "200",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response201": {
    "statusCode": "201",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response204": {
    "statusCode": "204",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response400": {
    "statusCode": "400",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response403": {
    "statusCode": "403",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response404": {
    "statusCode": "404",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response409": {
    "statusCode": "409",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response412": {
    "statusCode": "412",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response422": {
    "statusCode": "422",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response500": {
    "statusCode": "500",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },
  "response504": {
    "statusCode": "504",
    "responseParameters": "$${responseParameters}",
    "responseTemplates": "$${responseTemplates}",
    "responseModels": "$${responseModels}"
  },

  "responseParameters": {
    "method.response.header.Access-Control-Allow-Origin": "'*'"
  },
  "responseTemplates": { "application/json": "" },
  "responseModels": {}
}
```

At the s-function.json level:
```
{
  "functionName": "[NAME]",
  "endpointsTemplate": [
    "$${endpointsGetTemplate}", // handles both get and post - would really like multiple endpoint declarations per s-function.json to truly unleash this
    "$${endpointsPostTemplate}"
  ],
  "requestParametersTemplate": {
    [...]
  },
  "requestTemplatesTemplate": {
    "application/json": {
      [...]
  }
}
```

The entirety of s-function now looks like so:
```
"$${functionTemplate}"
```

(particularly @ac360) It seems wrong for s-function.json to end up so content-free because I want my API to abide by relative standards and require minimum verbosity and restatement.  This is why I've been going on about ideas of an inheritance and overriding scheme.  I think there's good room for doing away with all s-template.json files and using the s-[project/component/function].json files in a similar inheritance and override like fashion.  A configuration stack and merge mechanism would become necessary but it shouldn't be too rough to reasonably generally and clearly define those.

**Comment by erikerikson** [comment]
Finally getting around to some testing, I get an error related to the lack of broad template and variable definition reported here: https://github.com/serverless/serverless/issues/681

**Comment by staceymoore** [comment]
Latest iteration of my request template as an example for people who ask about it:

```
"requestTemplates": {
    "application/json": {
      "context": {
        "apiId": "$context.apiId",
        "authorizerPrincipalId": "$context.authorizer.principalId",
        "httpMethod": "$context.httpMethod",
        "identity": {
          "accountId": "$context.identity.accountId",
          "apiKey": "$context.identity.apiKey",
          "caller": "$context.identity.caller",
          "cognitoAuthenticationProvider": "$context.identity.cognitoAuthenticationProvider",
          "cognitoAuthenticationType": "$context.identity.cognitoAuthenticationType",
          "cognitoIdentityId": "$context.identity.cognitoIdentityId",
          "cognitoIdentityPoolId": "$context.identity.cognitoIdentityPoolId",
          "sourceIp": "$context.identity.sourceIp",
          "user": "$context.identity.user",
          "userAgent": "$context.identity.userAgent",
          "userArn": "$context.identity.userArn"
        },
        "lambda": "",
        "requestId": "$context.requestId",
        "resourceId": "$context.resourceId",
        "resourcePath": "$context.resourcePath",
        "stage": "$context.stage"
      },
      "header": {
        "accept": "$input.params().header.get('Accept')",
        "apiKey": "$input.params().header.get('x-api-key')",
        "authorization": "$input.params().header.get('Authorization')",
        "port": "$input.params().header.get('X-Forwarded-Port')",
        "protocol": "$input.params().header.get('X-Forwarded-Proto')",
        "sourceIp": "$input.params().header.get('X-Forwarded-For')"
      },
      "path": {
        "id": "$input.params().path.get('id')"
      },
      "querystring": {
        "id": "$input.params().querystring.get('id')"
      },
      "body": "$input.json('$')"
    }
  }
```

In the handler at runtime, I append the context as context.lamba to have access to the log stream name and aws request id

The s-function.json just has one line for the request template:
```
"requestTemplates": "$${requestTemplates}"
```

**Comment by goncaloneves** [comment]
I was trying to implement dynamic mapping for request path and query objects so I came to this solution.

Cheers for your work with Serverless. Great stuff :+1: 

```javascript
{
  "path": "{#foreach($key in $input.params().path.keySet())#if($foreach.index > 0),#end\"$key\":\"$input.params().path.get($key)\"#end}",
  "query": "{#foreach($key in $input.params().querystring.keySet())#if($foreach.index > 0),#end\"$key\":\"$input.params().querystring.get($key)\"#end}"
}
```

**Comment by erikerikson** [comment]
@goncaloneves there's a great version of this that provides method, headers, and the body here: http://kennbrodhagen.net/2015/12/06/how-to-create-a-request-object-for-your-lambda-event-from-api-gateway/
Template from that link:
```
{
  "body" : $input.json('$'),
  "headers": {
    #foreach($header in $input.params().header.keySet())
    "$header": "$util.escapeJavaScript($input.params().header.get($header))" #if($foreach.hasNext),#end
    #end
  },
  "method": "$context.httpMethod",
  "params": {
    #foreach($param in $input.params().path.keySet())
    "$param": "$util.escapeJavaScript($input.params().path.get($param))" #if($foreach.hasNext),#end
    #end
  },
  "query": {
    #foreach($queryParam in $input.params().querystring.keySet())
    "$queryParam": "$util.escapeJavaScript($input.params().querystring.get($queryParam))" #if($foreach.hasNext),#end
    #end
  }  
}
```

**Comment by goncaloneves** [comment]
@erikerikson this version one is a lot better ! Thank you.

**Comment by flomotlik** [comment]
@erikerikson @staceymoore @goncaloneves not sure if this issue is still in discussion? Sounds to me from the discussion that its basically done? I'll close for now, please open up if you still need it top be open.

**Comment by jokeyrhyme** [comment]
Hi folks! I'm pretty excited about `serverless`, but I'm a bit stuck here. Is there any documentation or example code for the error handling templates discussed here?

I've read through an upstream AWS blog post on this subject: https://aws.amazon.com/blogs/compute/error-handling-patterns-in-amazon-api-gateway-and-aws-lambda/

I'm trying to figure out how to return appropriate HTTP status codes with `serverless`, but I'm a little stuck.

Do I need to do anything to my project to get the templates in https://github.com/serverless/serverless/issues/608#issuecomment-184954698 to be predefined for me?

What should I do with the callback invocation in my function? I've tried adding a `"statusCode"` property with value `"500"` to my error object before passing it to the callback. I've tried passing a `JSON.stringify()`ed custom error object. I'm just a bit stuck with what to try next.

I've also tried a bunch of tinkering with Body Mapping Templates in API Gateway, and I noticed that none of the default templates mentioned in this thread seem to be visible there.
